erdos/src/communication/data_plane/stream_manager.rs

```diff
@@ -0,0 +1,373 @@
+        &mut self,
+        job: Job,
+        channel_to_data_sender: UnboundedSender<InterWorkerMessage>,
+    ) {
+        self.add_send_endpoint(
+            job,
+            SendEndpoint::InterProcess(self.stream_id, channel_to_data_sender),
+        );
+    }
+
+    fn add_inter_worker_recv_endpoint(&mut self, job: Job) -> Result<(), CommunicationError> {
+        let recv_endpoint_from_pusher = {
+            let mut pusher = self.pusher.lock().unwrap();
+            if let Some(pusher) = pusher.as_any().downcast_mut::<Pusher<Arc<Message<D>>>>() {
+                let (send_endpoint_to_pusher, recv_endpoint_from_pusher) =
+                    mpsc::unbounded_channel();
+                pusher.add_endpoint(job, SendEndpoint::InterThread(send_endpoint_to_pusher));
+                Some(RecvEndpoint::InterThread(recv_endpoint_from_pusher))
+            } else {
+                None
+            }
+        };
+
+        match recv_endpoint_from_pusher {
+            Some(recv_endpoint_from_pusher) => {
+                self.add_recv_endpoint(job, recv_endpoint_from_pusher);
+                Ok(())
+            }
+            None => Err(CommunicationError::StreamManagerError(format!(
+                "Error casting Pusher when adding inter-worker \
+                    receive endpoint for Stream {} (ID={}).",
+                self.stream_name, self.stream_id,
+            ))),
+        }
+    }
+
+    fn clone_pusher(&self) -> Arc<Mutex<dyn PusherT>> {
+        Arc::clone(&self.pusher)
+    }
+}
+
+/// A [`StreamManager`] is a data structure that stores the constructed
+/// endpoints for each stream. The [`DataPlane`] uses the [`StreamManager`]
+/// to initiate both Inter-Worker and Intra-Worker connections for all
+/// the [`Job`]s that are scheduled on the current [`Worker`]. Once the
+/// connections are set up, the [`Operator`]s cast the [`StreamEndpoints`]
+/// to the required message type and retrieve or send the data.
+pub(crate) struct StreamManager {
+    /// The [`Worker`] to which the [`StreamManager`] belongs.
+    worker_id: WorkerId,
+    /// Stores a `StreamEndpoints` for each stream id.
+    stream_endpoints: HashMap<StreamId, Box<dyn StreamEndpointsT>>,
+}
+
+impl StreamManager {
+    /// Initializes a new [`StreamManager`] for the [`Worker`] with the given ID.
+    pub fn new(worker_id: WorkerId) -> Self {
+        Self {
+            worker_id,
+            stream_endpoints: HashMap::new(),
+        }
+    }
+
+    /// Retrieves the ID of the [`Worker`] for which this [`StreamManager`] was created.
+    pub fn worker_id(&self) -> WorkerId {
+        self.worker_id
+    }
+
+    /// Adds an Intra-Worker send and receive endpoint for a [`WriteStream`].
+    /// The [`receiving_job`] specifies the [`Job`] that will construct a [`ReadStream`]
+    /// from the corresponding [`RecvEndpoint`].
+    pub fn add_intra_worker_endpoint(
+        &mut self,
+        stream: &Box<dyn AbstractStreamT>,
```

Comment:
```suggestion
        stream: &dyn AbstractStreamT,
```
Should work the same and provide more flexibility (since Box implements AsRef). I recommend making this change elsewhere in the file too.